addons/l10n_it_edi_doi/models/account_move.py

```diff
@@ -0,0 +1,228 @@
+                invoiced_amount=declaration_invoiced,
+            )
+
+            threshold_warning = declaration._build_threshold_warning_message(declaration_invoiced, declaration_not_yet_invoiced)
+
+            move.l10n_it_edi_doi_warning = '{}\n\n{}'.format('\n'.join(validity_warnings), threshold_warning).strip()
+
+    @api.depends('l10n_it_edi_doi_id')
+    def _compute_fiscal_position_id(self):
+        super()._compute_fiscal_position_id()
+        for move in self:
+            declaration_fiscal_position = move.company_id.l10n_it_edi_doi_fiscal_position_id
+            if declaration_fiscal_position and move.l10n_it_edi_doi_id:
+                move.fiscal_position_id = declaration_fiscal_position
+
+    def copy_data(self, default=None):
+        data_list = super().copy_data(default)
+        for move, data in zip(self, data_list):
+            declaration = move.l10n_it_edi_doi_id
+            date = fields.Date.context_today(self)
+            validity_warnings = declaration._get_validity_warnings(
+                move.company_id,
+                move.partner_id.commercial_partner_id,
+                move.currency_id,
+                date,
+                only_blocking=True,
+            )
+            if validity_warnings:
+                del data['l10n_it_edi_doi_id']
+                del data['fiscal_position_id']
+        return data_list
+
+    @api.constrains('l10n_it_edi_doi_id')
+    def _check_l10n_it_edi_doi_id(self):
+        for move in self:
+            declaration = move.l10n_it_edi_doi_id
+            if not declaration:
+                return
+            validity_errors = declaration._get_validity_errors(
+                move.company_id, move.partner_id.commercial_partner_id, move.currency_id
+            )
+            if validity_errors:
+                raise UserError('\n'.join(validity_errors))
+
+    def _post(self, soft=True):
+        errors = []
+        for move in self:
+            declaration_of_intent_tax = move.company_id.l10n_it_edi_doi_tax_id
+            if not declaration_of_intent_tax:
+                continue
+
+            declaration = move.l10n_it_edi_doi_id
+            if declaration:
+                partner = move.commercial_partner_id
+                date = move.l10n_it_edi_doi_date
+                validity_warnings = declaration._get_validity_warnings(
+                    move.company_id, partner, move.currency_id, date,
+                    invoiced_amount=move.l10n_it_edi_doi_amount,
+                    only_blocking=True
+                )
+                errors.extend(validity_warnings)
+
+            for line in move.invoice_line_ids:
+                if declaration_of_intent_tax not in line.tax_ids:
+                    continue
+                if not declaration:
+                    errors.append(_('Given the tax %s is applied, there should be a Declaration of Intent selected.',
+                                    declaration_of_intent_tax.name))
```

Comment:
This should not be in the `for` loop since it does not depend on `line` :wink:

Reply:
It depends on the line.
We only reach this `if` in case the special declaration tax is on the line
(due to the `if`/`continue` right before).
Should I rewrite the body of the `if`?